File: eco-core/core-plugin/src/main/java/com/willfp/ecoweapons/commands/CommandGive.java
package com.willfp.ecoweapons.commands;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.command.CommandHandler;
import com.willfp.eco.core.command.TabCompleteHandler;
import com.willfp.eco.core.command.impl.Subcommand;
import com.willfp.eco.core.config.updating.ConfigUpdater;
import com.willfp.ecoweapons.sets.Weapon;
import com.willfp.ecoweapons.sets.Weapons;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.util.StringUtil;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class CommandGive extends Subcommand {
    /**
     * The cached names.
     */
    private static final List<String> ITEM_NAMES = new ArrayList<>();

    /**
     * The cached numbers.
     */
    private static final List<String> NUMBERS = Arrays.asList(
            "1",
            "2",
            "3",
            "4",
            "5",
            "10",
            "32",
            "64"
    );

    /**
     * Instantiate a new command handler.
     *
     * @param plugin The plugin for the commands to listen for.
     */
    public CommandGive(@NotNull final EcoPlugin plugin) {
        super(plugin, "give", "ecoweapons.command.give", false);
        reload();
    }

    /**
     * Called on reload.
     */
    @ConfigUpdater
    public static void reload() {
        ITEM_NAMES.clear();
        ITEM_NAMES.addAll(Weapons.values().stream().map(Weapon::getName).collect(Collectors.toList()));
    }

    @Override
    public CommandHandler getHandler() {
        return (sender, args) -> {
            if (args.isEmpty()) {
                sender.sendMessage(this.getPlugin().getLangYml().getMessage("needs-player"));
                return;
            }

            if (args.size() == 1) {
                sender.sendMessage(this.getPlugin().getLangYml().getMessage("needs-item"));
                return;
            }

            String recieverName = args.get(0);
            Player reciever = Bukkit.getPlayer(recieverName);

            if (reciever == null) {
                sender.sendMessage(this.getPlugin().getLangYml().getMessage("invalid-player"));
                return;
            }

            String itemID = args.get(1);

            int amount = 1;

            Weapon weapon = Weapons.getByName(itemID.toLowerCase());

            if (weapon == null) {
                sender.sendMessage(this.getPlugin().getLangYml().getMessage("invalid-item"));
                return;
            }

            String message = this.getPlugin().getLangYml().getMessage("give-success");
            message = message.replace("%item%", weapon.getName()).replace("%recipient%", reciever.getName());
            sender.sendMessage(message);

            if (args.size() == 3) {
                try {
                    amount = Integer.parseInt(args.get(2));
                } catch (NumberFormatException ignored) {
                    // do nothing
                }
            }

            ItemStack item = weapon.getItem();
            item.setAmount(amount);
            reciever.getInventory().addItem(item);

        };
    }

    @Override
    public TabCompleteHandler getTabCompleter() {
        return (sender, args) -> {

            List<String> completions = new ArrayList<>();

            if (args.isEmpty()) {
                // Currently, this case is not ever reached
                return ITEM_NAMES;
            }

            if (args.size() == 1) {
                StringUtil.copyPartialMatches(args.get(0), Bukkit.getOnlinePlayers().stream().map(Player::getName).collect(Collectors.toList()), completions);
                return completions;
            }

            if (args.size() == 2) {
                StringUtil.copyPartialMatches(args.get(1), ITEM_NAMES, completions);

                Collections.sort(completions);
                return completions;
            }

            if (args.size() == 3) {
                StringUtil.copyPartialMatches(args.get(2), NUMBERS, completions);

                return completions;
            }

            return new ArrayList<>(0);
        };
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/ecoweapons/commands/CommandGive.kt
package com.willfp.ecoweapons.commands

import com.willfp.eco.core.EcoPlugin
import com.willfp.eco.core.command.CommandHandler
import com.willfp.eco.core.command.TabCompleteHandler
import com.willfp.eco.core.command.impl.Subcommand
import com.willfp.ecoweapons.weapons.Weapons
import org.bukkit.Bukkit
import org.bukkit.inventory.ItemStack
import org.bukkit.util.StringUtil

class CommandGive(plugin: EcoPlugin) : Subcommand(plugin, "give", "ecoweapons.command.give", false) {
    private val numbers = listOf(
        "1",
        "2",
        "3",
        "4",
        "5",
        "10",
        "32",
        "64"
    )

    override fun getHandler(): CommandHandler {
        return CommandHandler { sender, args ->
            if (args.isEmpty()) {
                sender.sendMessage(plugin.langYml.getMessage("needs-player"))
                return@CommandHandler
            }
            if (args.size == 1) {
                sender.sendMessage(plugin.langYml.getMessage("needs-item"))
                return@CommandHandler
            }
            val receiverName = args[0]
            val receiver = Bukkit.getPlayer(receiverName)
            if (receiver == null) {
                sender.sendMessage(plugin.langYml.getMessage("invalid-player"))
                return@CommandHandler
            }
            val itemID = args[1]
            var amount = 1
            val weapon = Weapons.getByID(itemID.lowercase().replace("_fuel", ""))
            if (weapon == null) {
                sender.sendMessage(plugin.langYml.getMessage("invalid-item"))
                return@CommandHandler
            }
            var message = plugin.langYml.getMessage("give-success")
            message = message.replace("%item%", weapon.id).replace("%recipient%", receiver.name)
            sender.sendMessage(message)
            if (args.size == 3) {
                amount = args[2].toIntOrNull() ?: 1
            }
            val item: ItemStack = if (itemID.contains("fuel")) weapon.fuelItem else weapon.itemStack
            item.amount = amount
            receiver.inventory.addItem(item)
        }
    }

    override fun getTabCompleter(): TabCompleteHandler {
        return TabCompleteHandler { _, args ->
            val completions = mutableListOf<String>()

            if (args.isEmpty()) {
                return@TabCompleteHandler Weapons.values().map { it.id }
            }

            if (args.size == 1) {
                StringUtil.copyPartialMatches(
                    args[0],
                    Bukkit.getOnlinePlayers().map { it.name },
                    completions)
                return@TabCompleteHandler completions
            }

            if (args.size == 2) {
                val weaponNames = Weapons.values().map { it.id } union
                        Weapons.values().mapNotNull { if (it.fuelEnabled) "${it.id}_fuel" else null }

                StringUtil.copyPartialMatches(args[1], weaponNames, completions)
                completions.sort()
                return@TabCompleteHandler completions
            }

            if (args.size == 3) {
                StringUtil.copyPartialMatches(args[2], numbers, completions)
                return@TabCompleteHandler completions
            }

            emptyList()
        }
    }
}
